bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -424,18 +409,32 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
   addTimers();
 }
 
-PreProcessor::~PreProcessor() {
-  LOG_TRACE(logger(), "~PreProcessor start");
+void PreProcessor::stop() {
+  LOG_DEBUG(logger(), "Going to stop PreProcessor");
+  lock_guard<mutex> lock(stopGuard_);
```

Comment:
If I understand correctly, stop() func is only called from the dtor, hence it is guaranteed to be called only from one context (assuming no double destruction of the preprocessor). Thus, the lock here is redundant. what actually solved the problem was adding  "if (!msgLoopDone_)" in PreProcessor::messageHandler, which prevents writing to the preprocessor msg queue during the stop function, which deletes msgs from the same queue. Am I missing something?

Reply:
That's the point. It is also called when the replica gets destructed (as a callback).

Reply:
So maybe the correct solution is to have the stop func called only from a single context? 
We do have a mix in our code base of RAII and "start-stop" patterns, but it seems like in the preprocessor case we use both approaches together...

Reply:
Actually, Tytus has already pushed the code that calls the stop() callback, I can remove the call from the dtor.

Reply:
I'll do that, thanks.